samples/subgroup-scan.cpp

```diff
@@ -0,0 +1,208 @@
+#include <numeric>
+#include <vector>
+
+// Dummy struct to generate unique kernel name types
+template <typename T, typename U, typename V>
+struct kernel_name {};
+
+/* Performs an inclusive scan with the given associative binary operation `Op`
+ * on the data in the `in` buffer. Runs in parallel on the provided accelerated
+ * hardware queue. Modifies the input buffer to contain the results of the scan.
+ * Input size has to be a power of two. If the size isn't so, the input can
+ * easily be padded to the nearest power of two with any values, and the scan on
+ * the meaningful part of the data will stay the same. */
+template <typename T, typename Op>
+void par_scan(sycl::buffer<T, 1> in, sycl::queue q) {
+  if ((in.get_count() & (in.get_count() - 1)) != 0 || in.get_count() == 0) {
+    throw std::runtime_error("Given input size is not a power of two.");
+  }
+
+  auto dev = q.get_device();
+
+  // Check if there is enough global memory.
+  size_t global_mem_size = dev.get_info<sycl::info::device::global_mem_size>();
+  if (!dev.is_host() && in.get_count() > (global_mem_size)) {
+    throw std::runtime_error("Input size exceeds device global memory size.");
+  }
+
+  // Obtain device limits.
+  size_t max_wgroup_size =
+      dev.get_info<sycl::info::device::max_work_group_size>();
+  size_t local_mem_size = dev.get_info<sycl::info::device::local_mem_size>();
+
+  /* Find a work-group size that is guaranteed to fit in local memory and is
+   * below the maximum work-group size of the device. */
+  size_t wgroup_size_lim =
+      sycl::min(max_wgroup_size, local_mem_size / sizeof(T));
+
+  size_t input_size = in.get_count();
+
+  size_t wgroup_size = 0;
+  for (size_t pow = size_t(1) << (sizeof(size_t) * 8 - 1); pow > 0; pow >>= 1) {
+    if ((input_size / pow) * pow == input_size && pow <= wgroup_size_lim) {
+      wgroup_size = pow;
+      break;
+    }
+  }
+
+  if (wgroup_size == 0) {
+    throw std::runtime_error(
+        "Could not find an appropriate work-group size for the given input.");
+  }
+
+  using namespace sycl::access;
+  q.submit([&](sycl::handler& cgh) {
+    auto data = in.template get_access<mode::read_write>(cgh);
+    auto temp = sycl::accessor<T, 1, mode::read_write, target::local>(
+        sycl::range<1>(wgroup_size), cgh);
+    cgh.parallel_for<kernel_name<T, Op, class scan_segments>>(
+        sycl::nd_range<1>(input_size, wgroup_size), [=](sycl::nd_item<1> item) {
+          size_t gid = item.get_global_linear_id();
+
+          auto sub_group = item.get_sub_group();
+          auto scan_res = inclusive_scan_over_group(sub_group, data[gid], Op{});
+          if (sub_group.get_local_id() == sub_group.get_local_range() - 1) {
+            temp[sub_group.get_group_id()[0]] = scan_res;
+          }
+          item.barrier(sycl::access::fence_space::local_space);
+          for (auto i = 1u; i < sub_group.get_group_range()[0]; i++) {
+            scan_res += sub_group.get_group_id()[0] >= i ? temp[i - 1] : 0;
```

Comment:
```suggestion
          for (auto i = 1u; i < sub_group.get_group_linear_range(); i++) {
            scan_res += sub_group.get_group_linear_id() >= i ? temp[i - 1] : 0;
```
Same goes for most spaces where you have `get_id/range()[0]`.

Reply:
Applied locally